[inductor][caching] load imc from disk

nmacchioni opened the pull request:
Summary: the reverse of dump

Test Plan: ```buck test fbcode//mode/opt caffe2/test/inductor:caching```

Differential Revision: D88984164




cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela

v0i0 commented on torch/_inductor/runtime/caching/config.py:
```diff
@@ -1,14 +1,33 @@
+
 @cache
 def _env_var_config(env_var: str, default: bool) -> bool:
-    if (env_val := os.environ.get(env_var)) is not None:
```
huh?

v0i0 commented on torch/_inductor/runtime/caching/config.py:
```diff
@@ -73,11 +92,43 @@ def _versioned_config(
+# IS_DUMP_MEMOIZER_CACHE_ENABLED is set to true.
+#
+# Set via environment variable: TORCHINDUCTOR_CACHE_DUMP_FILE_PATH=/path/to/dump.json
+_CACHE_DUMP_FILE_PATH_ENV_VAR: str = "TORCHINDUCTOR_CACHE_DUMP_FILE_PATH"
```
this is named as if it's the file you dump to, but it is the file we populate the cache from

nmacchioni committed:
[inductor][caching] Memoizer content loads from disk (#170237)

Summary:

This diff implements the reverse of the cache dump functionality - allowing Memoizers to pre-populate their in-memory caches by loading from disk dump files on initialization.

**Key Changes:**
* **New config option:** Added CACHE_DUMP_FILE_PATH environment variable to specify a dump file path for cache pre-population
* **Cache loading on init:** Memoizers now automatically load cache entries from the configured dump file during `__init__`
* **Hierarchical loading support:**
    * Standard Memoizers load from root-level cache entries
    * PersistentMemoizers with sub_key load from nested cache_entries[sub_key] structure
* **Robust error handling:** Gracefully handles missing files, corrupt JSON, and missing config
* **Enhanced _read_dump_from_disk():** Now accepts optional filepath parameter for reading arbitrary dump files

**Use Cases:**
* Pre-warming caches with known-good values across runs
* Reproducing behavior from previous executions
* Sharing cached computations across different environments
* Faster startup by avoiding recomputation

**Testing:**
Added comprehensive test coverage (7 new tests) covering:
* Basic cache loading from dump files
* Edge cases (missing files, corrupt JSON, unconfigured paths)
* Hierarchical loading for PersistentMemoizer with sub_keys
* End-to-end replay workflow using preloaded caches

Test Plan: ```buck test fbcode//mode/opt caffe2/test/inductor:caching```

Reviewed By: v0i0

Differential Revision: D88984164
